src/Decode.elm

```diff
@@ -151,3 +160,124 @@ unsafeToChar n =
+
+This tries to take the biggest step possible within a `Decode.loop` iteration.
+There is also some manual inlining to limit the number of function calls.
+
+-}
+decode18Bytes : Decode.Decoder String
+decode18Bytes =
+    Decode.map5 decode18Help
+        u32BE
+        u32BE
+        u32BE
+        u32BE
+        u16BE
+
+
+decode18Help : Int -> Int -> Int -> Int -> Int -> String
+decode18Help a b c d e =
+    let
+        combined1 =
+            Bitwise.shiftRightZfBy 8 a
+
+        combined2 =
+            Bitwise.or
+                (Bitwise.and 0xFF a |> Bitwise.shiftLeftBy 16)
+                (Bitwise.shiftRightZfBy 16 b)
+
+        combined3 =
+            Bitwise.or
+                (Bitwise.and 0xFFFF b |> Bitwise.shiftLeftBy 8)
+                (Bitwise.shiftRightZfBy 24 c)
+
+        combined4 =
+            Bitwise.and 0x00FFFFFF c
+
+        combined5 =
+            Bitwise.shiftRightZfBy 8 d
+
+        combined6 =
+            Bitwise.or
+                (Bitwise.and 0xFF d |> Bitwise.shiftLeftBy 16)
+                e
+    in
+    ""
+        |> bitsToCharsSpec combined6 combined5 combined4
+        |> bitsToCharsSpec combined3 combined2 combined1
+
+
+bitsToCharsSpec : Int -> Int -> Int -> String -> String
+bitsToCharsSpec bits1 bits2 bits3 accum =
+    -- Performance: prevent calls to `bitsToChar`. The overhead of function calls
+    -- became significant. This also allows more efficient string creation using String.cons
+    let
+        -- any 6-bit number is a valid base64 digit, so this is actually safe
+        p =
+            unsafeToChar (Bitwise.shiftRightZfBy 18 bits1)
+
+        q =
+            unsafeToChar (Bitwise.and (Bitwise.shiftRightZfBy 12 bits1) lowest6BitsMask)
+
+        r =
+            unsafeToChar (Bitwise.and (Bitwise.shiftRightZfBy 6 bits1) lowest6BitsMask)
+
+        s =
+            unsafeToChar (Bitwise.and bits1 lowest6BitsMask)
+    in
+    let
+        a =
+            unsafeToChar (Bitwise.shiftRightZfBy 18 bits2)
+
+        b =
+            unsafeToChar (Bitwise.and (Bitwise.shiftRightZfBy 12 bits2) lowest6BitsMask)
+
+        c =
+            unsafeToChar (Bitwise.and (Bitwise.shiftRightZfBy 6 bits2) lowest6BitsMask)
+
+        d =
+            unsafeToChar (Bitwise.and bits2 lowest6BitsMask)
+    in
+    let
+        x =
+            unsafeToChar (Bitwise.shiftRightZfBy 18 bits3)
+
+        y =
+            unsafeToChar (Bitwise.and (Bitwise.shiftRightZfBy 12 bits3) lowest6BitsMask)
+
+        z =
+            unsafeToChar (Bitwise.and (Bitwise.shiftRightZfBy 6 bits3) lowest6BitsMask)
+
+        w =
+            unsafeToChar (Bitwise.and bits3 lowest6BitsMask)
+    in
+    accum
+        |> String.cons s
+        |> String.cons r
+        |> String.cons q
+        |> String.cons p
+        |> String.cons d
+        |> String.cons c
+        |> String.cons b
+        |> String.cons a
+        |> String.cons w
+        |> String.cons z
+        |> String.cons y
+        |> String.cons x
```

Comment:
Most likely a list is significantly worse. Creating a list still has quite a memory cost. For instance `1 + 2 + 3` is much, much faster than `List.sum [1,2,3]`.

But, a lot of function calls happen here, so I'll try and get some hard numbers. It might be that the cost of function calls outweighs list creation/folding here. But I suspect not.

Reply:
right. I'd put this in an ellie but it still runs 0.19.1 (elm-benchmark triggers the Map.! bug) and this should be ran with `--optimize` anyway. 

Results: using `String.cons` is roughly 2.5 times faster here 

```
old: 1,156,170
new: 3,866,828
change: +234.45%
```

Normally this wouldn't matter of course, but `fromBytes` is constrained by two factors: how fast can the input be read and the output be written. So in this case it does really matter than string creation is fast. I'll link to this comment in the code.

```elm
module Main exposing (main)

import Benchmark exposing (..)
import Benchmark.Runner exposing (BenchmarkProgram, program)


main : BenchmarkProgram
main =
    program suite


suite : Benchmark
suite =
    comp


comp =
    let
        s =
            'a'

        r =
            'a'

        q =
            'a'

        p =
            'a'

        d =
            'a'

        c =
            'a'

        b =
            'a'

        a =
            'a'

        w =
            'a'

        z =
            'a'

        y =
            'a'

        x =
            'a'
    in
    Benchmark.compare "string from chars"
        "old"
        (\_ ->
            String.fromList [ s, r, q, p, d, c, b, a, w, z, y, x ]
        )
        "new"
        (\_ ->
            ""
                |> String.cons s
                |> String.cons r
                |> String.cons q
                |> String.cons p
                |> String.cons d
                |> String.cons c
                |> String.cons b
                |> String.cons a
                |> String.cons w
                |> String.cons z
                |> String.cons y
                |> String.cons x
        )
```